patches/0004-Output-to-hashed-mappings.patch

```diff
@@ -0,0 +1,1622 @@
++import net.fabricmc.loom.util.ModRemappingHelper;
+ import net.fabricmc.loom.util.TinyRemapperHelper;
+ import net.fabricmc.loom.util.ZipUtils;
+ import net.fabricmc.loom.util.kotlin.KotlinClasspathService;
+@@ -134,7 +135,7 @@ public class ModProcessor {
+ 		AccessWidenerRemapper awRemapper = new AccessWidenerRemapper(
+ 				writer,
+ 				remapper,
+-				MappingsNamespace.INTERMEDIARY.toString(),
++				AccessWidenerReader.readHeader(input).getNamespace(),
+ 				MappingsNamespace.NAMED.toString()
+ 		);
+ 		AccessWidenerReader reader = new AccessWidenerReader(awRemapper);
+@@ -148,10 +149,20 @@ public class ModProcessor {
+ 		Path[] mcDeps = project.getConfigurations().getByName(Constants.Configurations.LOADER_DEPENDENCIES).getFiles()
+ 				.stream().map(File::toPath).toArray(Path[]::new);
+ 
+-		project.getLogger().lifecycle(":remapping " + remapList.size() + " mods (TinyRemapper, " + fromM + " -> " + toM + ")");
++		final Map<ModDependencyInfo, MappingsNamespace> intermediateMap = new HashMap<>();
++
++		for (ModDependencyInfo info : remapList) {
++			intermediateMap.put(info, ModRemappingHelper.getIntermediateMappings(info.inputFile));
++		}
+ 
+-		TinyRemapper.Builder builder = TinyRemapper.newRemapper()
+-				.withMappings(TinyRemapperHelper.create(mappingsProvider.getMappings(), fromM, toM, false))
++		project.getLogger().lifecycle(":remapping " + remapList.size() + " mods");
++
++		TinyRemapper.Builder hashedBuilder = TinyRemapper.newRemapper()
++				.withMappings(TinyRemapperHelper.create(mappingsProvider.getMappings(), MappingsNamespace.HASHED.toString(), toM, false))
++				.renameInvalidLocals(false);
++
++		TinyRemapper.Builder intermediaryBuilder = TinyRemapper.newRemapper()
++				.withMappings(TinyRemapperHelper.create(mappingsProvider.getMappings(), MappingsNamespace.INTERMEDIARY.toString(), toM, false))
+ 				.renameInvalidLocals(false);
+ 
+ 		final KotlinClasspathService kotlinClasspathService = KotlinClasspathService.getOrCreateIfRequired(project);
+@@ -159,16 +170,22 @@ public class ModProcessor {
+ 
+ 		if (kotlinClasspathService != null) {
+ 			kotlinRemapperClassloader = KotlinRemapperClassloader.create(kotlinClasspathService);
+-			builder.extension(kotlinRemapperClassloader.getTinyRemapperExtension());
++			hashedBuilder.extension(kotlinRemapperClassloader.getTinyRemapperExtension());
++			intermediaryBuilder.extension(kotlinRemapperClassloader.getTinyRemapperExtension());
+ 		}
+ 
+-		final TinyRemapper remapper = builder.build();
++		final TinyRemapper hashedRemapper = hashedBuilder.build();
++		final TinyRemapper intermediaryRemapper = intermediaryBuilder.build();
++
++		final Map<MappingsNamespace, TinyRemapper> remapperMap = new HashMap<>();
++		remapperMap.put(MappingsNamespace.HASHED, hashedRemapper);
++		remapperMap.put(MappingsNamespace.INTERMEDIARY, intermediaryRemapper);
+ 
+-		for (Path minecraftJar : extension.getMinecraftJars(MappingsNamespace.INTERMEDIARY)) {
+-			remapper.readClassPathAsync(minecraftJar);
```

Comment:
It looks like you never fill the intermediary remapper's classpath?

Reply:
Hmm, there was a reason I did this but I don't remember it, you're right it would probably make more sense to populate the remapper's classpath.